flatland/envs/rewards.py

```diff
@@ -142,10 +142,17 @@ def __init__(self,
     def step_reward(self, agent: EnvAgent, agent_transition_data: AgentTransitionData, distance_map: DistanceMap, elapsed_steps: int) -> float:
         reward = 0
         if agent.position is not None:
-            self.states[agent.handle][Waypoint(agent.position, agent.direction)].add(agent.state)
-        if agent.position not in self.arrivals[agent.handle]:
-            self.arrivals[agent.handle][Waypoint(agent.position, agent.direction)] = elapsed_steps
-            self.departures[agent.handle][Waypoint(agent.old_position, agent.old_direction)] = elapsed_steps
+            wp = Waypoint(agent.position, agent.direction)
+            self.states[agent.handle][wp].add(agent.state)
+
+            # Only record arrival if this is a new waypoint (not dwelling at same position)
+            old_wp = Waypoint(agent.old_position, agent.old_direction)
+            if wp not in self.arrivals[agent.handle]:
+                self.arrivals[agent.handle][wp] = elapsed_steps
+                # Only record departure from old position if we actually moved
+                if old_wp != wp:
+                    self.departures[agent.handle][old_wp] = elapsed_steps
```

Comment:
I suggest the following simplification:
```suggestion
        wp = Waypoint(agent.position, agent.direction)
        if agent.position is not None:
            self.states[agent.handle][wp].add(agent.state)
        if wp not in self.arrivals[agent.handle]:
            self.arrivals[agent.handle][wp] = elapsed_steps
            old_wp = Waypoint(agent.old_position, agent.old_direction)
            if old_wp != wp:
                self.departures[agent.handle][old_wp] = elapsed_steps
```

Reply:
I do not completely get this - maybe because of the formatting / indentation. IMHO your suggestion has some drawbacks:

1. Creates Waypoint with None - might be invalid or cause unexpected behavior
2. Records arrivals/departures for None positions

Reply:
Both points valid 👍 Let's merge your version for now. 

One point though: Why is the `if old_wp != wp:` guard necessary? Should not arrival at new and departure at old always happen together, so why `wp not in self.arrivals[agent.handle]` is not enough? See comment in `test_departure_only_when_moving`